pkg/nsx/services/firewall.go

```diff
@@ -201,45 +218,77 @@ func (service *SecurityPolicyService) buildBasicTags(obj *v1alpha1.SecurityPolic
-		"Tag", "EQUALS",
+		"Tag", "EQUALS", "EQUALS",
 	)
 	expressions.Add(clusterExpression)
+	var err error = nil
+	var podSelectorCount, vMSelectorCount = 0, 0
+
 	if target.PodSelector != nil {
 		service.addOperatorIfNeeded(expressions, "AND")
 		// TODO: consider to use project_uid instead of project
 		nsExpression := service.buildExpression(
 			"Condition", "SegmentPort",
 			fmt.Sprintf("%s|%s", util.TagScopeNCPProject, obj.ObjectMeta.Namespace),
-			"Tag", "EQUALS",
+			"Tag", "EQUALS", "EQUALS",
 		)
 		expressions.Add(nsExpression)
-		service.updatePortExpressions(target.PodSelector.MatchLabels, expressions)
+		service.updateExpressionsMatchLables(target.PodSelector.MatchLabels, "SegmentPort", expressions)
+		if len(target.PodSelector.MatchLabels) > 0 {
+			podSelectorCount = 1
+		}
+
+		if target.PodSelector.MatchExpressions != nil {
+			mergedMatchExpressions := service.mergeSelectorMatchExpression(target.PodSelector.MatchExpressions)
+			podSelectorCount, err = service.validateSelectorExpressions(target.PodSelector.MatchLabels, *mergedMatchExpressions, "SegmentPort")
+			if err == nil {
+				service.updateExpressionsMatchExpression(*mergedMatchExpressions, target.PodSelector.MatchLabels,
+					&group.Expression, clusterExpression, nsExpression, "SegmentPort", expressions)
+			}
+		}
 	}
 	if target.VMSelector != nil {
 		service.addOperatorIfNeeded(expressions, "AND")
 		nsExpression := service.buildExpression(
 			"Condition", "SegmentPort",
 			fmt.Sprintf("%s|%s", util.TagScopeNCPVIFProject, obj.ObjectMeta.Namespace),
-			"Tag", "EQUALS",
+			"Tag", "EQUALS", "EQUALS",
 		)
 		expressions.Add(nsExpression)
-		service.updatePortExpressions(target.VMSelector.MatchLabels, expressions)
+		service.updateExpressionsMatchLables(target.VMSelector.MatchLabels, "SegmentPort", expressions)
+		if len(target.VMSelector.MatchLabels) > 0 {
+			vMSelectorCount = 1
+		}
+
+		if target.VMSelector.MatchExpressions != nil {
+			mergedMatchExpressions := service.mergeSelectorMatchExpression(target.VMSelector.MatchExpressions)
+			vMSelectorCount, err = service.validateSelectorExpressions(target.VMSelector.MatchLabels, *mergedMatchExpressions, "SegmentPort")
+			if err == nil {
+				service.updateExpressionsMatchExpression(*mergedMatchExpressions, target.VMSelector.MatchLabels,
+					&group.Expression, clusterExpression, nsExpression, "SegmentPort", expressions)
+			}
+		}
 	}
+
+	// Check Selector target group limitation
+	if podSelectorCount > MaxGroupExpressions {
+		errorMsg := fmt.Sprintf("Total number of PodSelector target expressions:%d exceed NSX limit of %d", podSelectorCount, MaxGroupExpressions)
+		err = errors.New(errorMsg)
+		log.Error(err, "updateTargetExpressions")
+	}
+	if vMSelectorCount > MaxGroupExpressions {
+		errorMsg := fmt.Sprintf("Total number of VMSelector target expressions:%d exceed NSX limit of %d", vMSelectorCount, MaxGroupExpressions)
+		err = errors.New(errorMsg)
+		log.Error(err, "updateTargetExpressions")
+	}
```

Comment:
I don't think we need check the count separately here, as you will check the total count outside

Reply:
The reason why we checked here is for one group limit. In fact, we have two limits:
1. Total number of groups, i.e. total max criteria in one NSGroup: 5
2. Maximum 5 expressions are allowed inside a nested expression
Here，we are checking the second one, inside one group, only allow 5 expressions. Outside, we need to check total 
group number limit.